taxcalc/simpletaxio.py

```diff
@@ -0,0 +1,648 @@
+                self._input[lnum] = vardict
+
+    def _validate_input(self):
+        """
+        Validate INPUT variable values stored in self._input dictionary.
+
+        Parameters
+        ----------
+        none: void
+
+        Raises
+        ------
+        ValueError:
+            if INPUT variable values are not in proper range of values.
+
+        Returns
+        -------
+        nothing: void
+
+        Notes
+        -----
+        The integer value of each input variable is stored in the
+        self._input dictionary, which is indexed by INPUT file line
+        number and by INPUT file variable number (where both indexes
+        begin with one).
+        """
+        self._year_set = set()
+        min_year = self._params.start_year
+        max_year = self._params.end_year
+        for lnum in range(1, len(self._input) + 1):
+            var = self._input[lnum]
+            year = var[2]
+            if year < min_year or year > max_year:
+                msg = ('var[2]={} on line {} of simtax INPUT is not in '
+                       '[{},{}] Parameters start-year, end-year range')
+                raise ValueError(msg.format(year, lnum, min_year, max_year))
+            if year not in self._year_set:
+                self._year_set.add(year)
+            state = var[3]
+            if state != 0:
+                msg = ('var[3]={} on line {} of simtax INPUT is not zero '
+                       'to indicate no state income tax calculations')
+                raise ValueError(msg.format(state, lnum))
+            filing_status = var[4]
+            if filing_status < 1 or filing_status > 3:
+                msg = ('var[4]={} on line {} of simtax INPUT is not '
+                       'in [1,3] filing-status range')
+                raise ValueError(msg.format(filing_status, lnum))
+            num_all_dependents = var[5]
+            if filing_status == 3 and num_all_dependents == 0:
+                msg = ('var[5]={} on line {} of simtax INPUT is not '
+                       'positive when var[4] equals 3')
+                raise ValueError(msg.format(num_all_dependents, lnum))
+            num_aged = var[6]
+            if filing_status == 2:
+                if num_aged > 2:
+                    msg = ('var[6]={} on line {} of simtax INPUT is not '
+                           'less than or equal to two')
+                    raise ValueError(msg.format(num_aged, lnum))
+            else:  # if filing_status is 1=single or 3=head_of_household
+                if num_aged > 1:
+                    msg = ('var[6]={} on line {} of simtax INPUT is not '
+                           'less than or equal to one')
+                    raise ValueError(msg.format(num_aged, lnum))
+            tran = var[13]
+            if tran != 0:
+                msg = ('var[13]={} on line {} of simtax INPUT is not zero '
+                       'to indicate no state income tax calculations')
```

Comment:
I can't be sure, but it looks to me that, based on the pattern from the other error checking in this function, the intention is to raise an exception if `tran !=0`. But after the `msg` variable is created, we leave this `for` block. Are you sure you did not want to raise a ValueError here?